### ISSUE TITLE
Project overview

### ISSUE THREAD

**Issue by kiootic** [oursky/gitlint-bot#10]
- The bot can be configured to watch for:
    - repositories within specific GitHub organizations; and
    - specific GitHub repositories.
- Periodically (configurable), the commits in configured repositories
  would be analyzed for:
    - commit message quality (linting)
- The analyzed result of commits should be retrievable for subsequent lookup.
- Each commit would be given a score, indicating how bad it is.
- At the end of collection period, the analyzed commits would be summarized as:
    - the top 5 repository with highest total commit score.
    - the top 5 committer with highest total commit score.
    - the top 10 commit with highest commit score.
- The summary would be delivered as a message to a configured Slack channel.

**Comment by Arthelon** [comment]
- Could include positive/negative scoring to rules to reward certain commit features (i.e. signed commit footers would yield a positive score) instead of only counting errors
- Instead of using the total commit scores in our summary message, I think a better measure to use would be average score per commit since repos/users with the most commits would have the highest/lower scores.
- Scoring summary might look something like:
    - Top 5 repositories with highest average score per commit
    - Top 5 users with highest average score per commit
    - Top 10 users with highest overall score

**Comment by kiootic** [comment]
> Could include positive/negative scoring to rules to reward certain commit features (i.e. signed commit footers would yield a positive score) instead of only counting errors

That's an interesting idea. However, positive scores and negative scores should be accounted for separately: e.g. a signed commit (positive) cannot cover for a bad commit message (negative)

> Instead of using the total commit scores in our summary message, I think a better measure to use would be average score per commit since repos/users with the most commits would have the highest/lower scores.

Agreed.

**Comment by Arthelon** [comment]
> However, positive scores and negative scores should be accounted for separately

I agree with this. Maybe we can add a field to distinguish between a positive/negative commit diagnosis later on. To start off, the rules should look for bad commit features and flag a commit is "not valid" if the rule check fails (i.e. length of header less than 50). 

I'm thinking a score is awarded for the a rule only if the commit diagnosis is valid (invalid commit diagnoses receive 0 score).

**Comment by Arthelon** [comment]
Also, in my data-flow diagram. The second process mentions periodically getting all the commits/diagnoses for a certain period of time and calculating the scores then. 

I think it makes more sense to attach the score to each commit right after the commit is linted.

**Comment by kiootic** [comment]
> I'm thinking a score is awarded for the a rule only if the commit diagnosis is valid (invalid commit diagnoses receive 0 score).

I'm not sure what's the relationship between commit diagnosis and validity. Can you explain more?

> I think it makes more sense to attach the score to each commit right after the commit is linted.

I agree. You may want to update the DFD to reflect it.

**Comment by Arthelon** [comment]
I was thinking that each commit diagnosis would contain the results of checking (or diagnosing) a commit message against one rule. An example of valid check would be for the `header-max-length` rule. If the header length is less than 50 then diagnosis is `valid`.

**Comment by kiootic** [comment]
Commit diagnosis should be emitted only if it's useful: a commit conforming to a rule emits no diagnosis, but a commit violating a rule would emit a diagnosis, with details on the violation. Therefore, there is no need to have 'validity' in commit diagnosis.

**Comment by Arthelon** [comment]
yeah, my bad. I was referring to commit diagnosis and rule checks as the same thing. I'll distinguish the two.
